galaxy_ng/contrib/metrics_middleware.py

```diff
@@ -0,0 +1,50 @@
+import base64
+import json
+
+from django_prometheus.middleware import (
+    Metrics,
+    PrometheusAfterMiddleware,
+    PrometheusBeforeMiddleware,
+)
+
+from galaxy_ng.app.auth.auth import RHIdentityAuthentication
+
+
+EXTENDED_METRICS = [
+    "django_http_requests_latency_seconds_by_view_method",
+    "django_http_responses_total_by_status_view_method",
+    "django_http_requests_total_by_view_transport_method",
+]
+
+
+class AccountEnhancedMetrics(Metrics):
+    def register_metric(self, metric_cls, name, documentation, labelnames=(), **kwargs):
+        if name in EXTENDED_METRICS:
+            labelnames += ("account",)
+        return super().register_metric(
+            metric_cls, name, documentation, labelnames=labelnames, **kwargs
+        )
+
+
+class AccountEnhancedMetricsBeforeMiddleware(PrometheusBeforeMiddleware):
+    metrics_cls = AccountEnhancedMetrics
+
+
+class AccountEnhancedMetricsAfterMiddleware(PrometheusAfterMiddleware):
+    metrics_cls = AccountEnhancedMetrics
+
+    def label_metric(self, metric, request, response=None, **labels):
+        new_labels = labels
+        if metric._name in EXTENDED_METRICS:
+            account = "unknown"
+            encoded_identity = request.META.get(RHIdentityAuthentication.header)
+            if encoded_identity is not None:
+                try:
+                    json_string = base64.b64decode(encoded_identity)
+                    identity_header = json.loads(json_string)
+                    account = identity_header.get("identity", {}).get("account_number", "unknown")
+                except ValueError:
+                    pass
+            new_labels = {"account": account}
+            new_labels.update(labels)
```

Comment:
Is adding the account number to the metric labels going to cause any issues?  Account number will be fairly high cardinality and wondering if that could become a problem.

https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels

> As a general guideline, try to keep the cardinality of your metrics below 10, and for metrics that exceed that, aim to limit them to a handful across your whole system. The vast majority of your metrics should have no labels.
> 
> If you have a metric that has a cardinality over 100 or the potential to grow that large, investigate alternate solutions such as reducing the number of dimensions or moving the analysis away from monitoring and to a general-purpose processing system.

Reply:
Agreed, considering number of accounts may be large and grow over time and that this code adds labels to all metrics, I'm concerned it may cause performance issues. Also this particular use case seems to be more appropriate for the logging system rather then metrics collection.